ICE caused by incorrect inference for unsuffixed int literal in final expression of function

```
fn foo() -> isize {
    0
}
```

Will ICE : 
```
.../gcc/testsuite/rust.test/execute/empty_main.rs:7:1: warning: unused name
.../gcc/testsuite/rust.test/execute/empty_main.rs: In function ‘foo’:
.../gcc/testsuite/rust.test/execute/empty_main.rs:7:1: error: non-trivial conversion in ‘integer_cst’
isize
i32
D.198 = 0;
/some/path//gccrs/gcc/testsuite/rust.test/execute/empty_main.rs:7:1: internal compiler error: ‘verify_gimple’ failed
0x1a5d602 verify_gimple_in_seq(gimple*)
        ../../gcc/tree-cfg.c:5119
0x1488836 gimplify_body(tree_node*, bool)
        ../../gcc/gimplify.c:15329
0x1488abe gimplify_function_tree(tree_node*)
        ../../gcc/gimplify.c:15400
0x115e2aa cgraph_node::analyze()
        ../../gcc/cgraphunit.c:670
0x11602f3 analyze_functions
        ../../gcc/cgraphunit.c:1235
0x116353f symbol_table::finalize_compilation_unit()
        ../../gcc/cgraphunit.c:2513
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
```

After a quick gdb session, it seems that conversion from the 0 literal (SI) to `isize` (DI) is needed  and makes the compiler crash.

The falling test is :
```
│B+>4425          if (!useless_type_conversion_p (lhs_type, rhs1_type))   
│   4426            {                                                                                         
│   4427              error ("non-trivial conversion in %qs", code_name);        
│   4428              debug_generic_expr (lhs_type);                           
│   4429              debug_generic_expr (rhs1_type);  
│   4430              return true;        
│   4431            } 
```
With the following arguments:
```
 <integer_type 0x7ffff7472738 isize sizes-gimplified DI
    size <integer_cst 0x7ffff745dbb8 type <integer_type 0x7ffff74720a8 bitsizetype> constant 64>
    unit-size <integer_cst 0x7ffff745dbd0 type <integer_type 0x7ffff7472000 sizetype> constant 8>
    align:64 warn_if_not_align:0 symtab:0 alias-set -1 canonical-type 0x7ffff7472738 precision:64 min <integer_cst 0x7ffff745de40 -9223372036854775808> max <integer_cst 0x7ffff745de58 9223372036854775807>>
$2 = void

 <integer_type 0x7ffff74725e8 i32 public SI
    size <integer_cst 0x7ffff745ddf8 type <integer_type 0x7ffff74720a8 bitsizetype> constant 32>
    unit-size <integer_cst 0x7ffff745de10 type <integer_type 0x7ffff7472000 sizetype> constant 4>
    align:32 warn_if_not_align:0 symtab:0 alias-set -1 canonical-type 0x7ffff74725e8 precision:32 min <integer_cst 0x7ffff745ddb0 -2147483648> max <integer_cst 0x7ffff745ddc8 2147483647>
    pointer_to_this <pointer_type 0x7ffff74799d8>>
$3 = void
```
This is rejected later on by :
```
│  67          bool  
│   68          useless_type_conversion_p (tree outer_type, tree inner_type)   
...
│   93            /* Changes in machine mode are never useless conversions because the RTL  
│   94               middle-end expects explicit conversions between modes.  */         
│  >95            if (TYPE_MODE (inner_type) != TYPE_MODE (outer_type))          
│   96              return false;
```

The full tree for the function:
```
 <function_decl 0x7ffff748f200 foo
    type <function_type 0x7ffff748e738
        type <integer_type 0x7ffff7472738 isize sizes-gimplified DI
            size <integer_cst 0x7ffff745dbb8 constant 64>
            unit-size <integer_cst 0x7ffff745dbd0 constant 8>
            align:64 warn_if_not_align:0 symtab:0 alias-set -1 canonical-type 0x7ffff7472738 precision:64 min <integer_cst 0x7ffff745de40 -9223372036854775808> max <integer_cst 0x7ffff745de58 9223372036854775807>>
        QI
        size <integer_cst 0x7ffff745dca8 constant 8>
        unit-size <integer_cst 0x7ffff745dcc0 constant 1>
        align:8 warn_if_not_align:0 symtab:0 alias-set -1 canonical-type 0x7ffff748e738
        arg-types <tree_list 0x7ffff7471b18 value <void_type 0x7ffff7472f18>>
        pointer_to_this <pointer_type 0x7ffff748e7e0>>
    public uninlinable QI /some/path//gccrs/gcc/testsuite/rust.test/execute/empty_main.rs:7:1 align:8 warn_if_not_align:0 initial <block 0x7ffff7493000>
    result <result_decl 0x7ffff7467168 D.196 type <integer_type 0x7ffff7472738 isize>
        ignored DI /some/path//gccrs/gcc/testsuite/rust.test/execute/empty_main.rs:7:1 size <integer_cst 0x7ffff745dbb8 64> unit-size <integer_cst 0x7ffff745dbd0 8>
        align:64 warn_if_not_align:0 context <function_decl 0x7ffff748f200 foo>>
    struct-function 0x7ffff7494000>
```

I'll continue and try to debug this on my own to better understand how this part works...

--- a/gcc/rust/typecheck/rust-hir-type-check-item.h
+++ b/gcc/rust/typecheck/rust-hir-type-check-item.h
@@ -77,15 +77,12 @@ public:
     auto expected_ret_tyty = resolve_fn_type->get_return_type ();
     context->push_return_type (expected_ret_tyty);
 
-    auto result
+    auto block_expr_ty
       = TypeCheckExpr::Resolve (function.get_definition ().get (), false);
-    auto ret_resolved = expected_ret_tyty->unify (result);
-    if (ret_resolved == nullptr)
-      return;
-
-    context->peek_return_type ()->append_reference (ret_resolved->get_ref ());
 
     context->pop_return_type ();
+
+    expected_ret_tyty->unify (block_expr_ty);
   }
 
 private:

--- a/gcc/rust/typecheck/rust-tyty-rules.h
+++ b/gcc/rust/typecheck/rust-tyty-rules.h
@@ -67,41 +67,39 @@ public:
       }
 
     other->accept_vis (*this);
-    if (resolved != nullptr)
+    if (resolved == nullptr)
+      return nullptr;
+
+    resolved->append_reference (get_base ()->get_ref ());
+    resolved->append_reference (other->get_ref ());
+    for (auto ref : get_base ()->get_combined_refs ())
+      resolved->append_reference (ref);
+    for (auto ref : other->get_combined_refs ())
+      resolved->append_reference (ref);
+
+    bool result_resolved = resolved->get_kind () != TyTy::TypeKind::INFER;
+    bool result_is_infer_var = resolved->get_kind () == TyTy::TypeKind::INFER;
+    bool results_is_non_general_infer_var
+      = (result_is_infer_var
+	 && ((InferType *) resolved)->get_infer_kind ()
+	      != TyTy::InferType::GENERAL);
+    if (result_resolved || results_is_non_general_infer_var)
       {
-	resolved->append_reference (get_base ()->get_ref ());
-	resolved->append_reference (other->get_ref ());
-	for (auto ref : get_base ()->get_combined_refs ())
-	  resolved->append_reference (ref);
-	for (auto ref : other->get_combined_refs ())
-	  resolved->append_reference (ref);
-
-	bool result_resolved = resolved->get_kind () != TyTy::TypeKind::INFER;
-	bool result_is_infer_var
-	  = resolved->get_kind () == TyTy::TypeKind::INFER;
-	bool results_is_non_general_infer_var
-	  = (result_is_infer_var
-	     && ((InferType *) resolved)->get_infer_kind ()
-		  != TyTy::InferType::GENERAL);
-	if (result_resolved || results_is_non_general_infer_var)
+	for (auto &ref : resolved->get_combined_refs ())
 	  {
-	    for (auto &ref : resolved->get_combined_refs ())
-	      {
-		TyTy::BaseType *ref_tyty = nullptr;
-		bool ok = context->lookup_type (ref, &ref_tyty);
-		if (!ok)
-		  continue;
+	    TyTy::BaseType *ref_tyty = nullptr;
+	    bool ok = context->lookup_type (ref, &ref_tyty);
+	    if (!ok)
+	      continue;
 
-		// if any of the types are inference variables lets fix them
-		if (ref_tyty->get_kind () == TyTy::TypeKind::INFER)
-		  {
-		    NodeId ref_node_id = UNKNOWN_NODEID;
-		    context->insert_type (
-		      Analysis::NodeMapping (mappings->get_current_crate (),
-					     ref_node_id, ref,
-					     UNKNOWN_LOCAL_DEFID),
-		      resolved->clone ());
-		  }
+	    // if any of the types are inference variables lets fix them
+	    if (ref_tyty->get_kind () == TyTy::TypeKind::INFER)
+	      {
+		context->insert_type (
+		  Analysis::NodeMapping (mappings->get_current_crate (),
+					 UNKNOWN_NODEID, ref,
+					 UNKNOWN_LOCAL_DEFID),
+		  resolved->clone ());
 	      }
 	  }
       }

--- a/gcc/rust/typecheck/rust-tyty.h
+++ b/gcc/rust/typecheck/rust-tyty.h
@@ -101,6 +101,24 @@ public:
 
   virtual bool has_subsititions_defined () const { return false; }
 
+  std::string mappings_str () const
+  {
+    std::string buffer = "Ref: " + std::to_string (get_ref ())
+			 + " TyRef: " + std::to_string (get_ty_ref ());
+    buffer += "[";
+    for (auto &ref : combined)
+      buffer += std::to_string (ref) + ",";
+    buffer += "]";
+    return "(" + buffer + ")";
+  }
+
+  std::string debug_str () const
+  {
+    return as_string () + ":" + mappings_str ();
+  }
+
+  void debug () const { printf ("%s\n", debug_str ().c_str ()); }
+
 protected:
   BaseType (HirId ref, HirId ty_ref, TypeKind kind,
 	    std::set<HirId> refs = std::set<HirId> ())

--- a/gcc/rust/typecheck/rust-tyty.h
+++ b/gcc/rust/typecheck/rust-tyty.h
@@ -57,7 +57,12 @@ public:
 
   HirId get_ref () const { return ref; }
 
-  void set_ref (HirId id) { ref = id; }
+  void set_ref (HirId id)
+  {
+    if (id != ref)
+      append_reference (ref);
+    ref = id;
+  }
 
   HirId get_ty_ref () const { return ty_ref; }
 

--- a/gcc/testsuite/rust.test/compilable/block_expr4.rs
+++ b/gcc/testsuite/rust.test/compilable/block_expr4.rs
@@ -1,0 +1,7 @@
+fn foo() -> isize {
+    0
+}
+
+fn main() {
+    let a = foo();
+}

